cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -70,43 +83,222 @@ func (o *checkCapacityProvClass) Provision(
 	daemonSets []*appsv1.DaemonSet,
 	nodeInfos map[string]*schedulerframework.NodeInfo,
 ) (*status.ScaleUpStatus, errors.AutoscalerError) {
-	if len(unschedulablePods) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
+	combinedStatus := NewCombinedStatusSet()
+	provisioningRequestsProcessed := 0
+	startTime := time.Now()
 
-	prs := provreqclient.ProvisioningRequestsForPods(o.client, unschedulablePods)
-	prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
-	if len(prs) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
-	// Pick 1 ProvisioningRequest.
-	pr := prs[0]
 	o.context.ClusterSnapshot.Fork()
 	defer o.context.ClusterSnapshot.Revert()
 
-	scaleUpIsSuccessful, err := o.checkcapacity(unschedulablePods, pr)
-	if err != nil {
-		return status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
-	}
-	if scaleUpIsSuccessful {
-		return &status.ScaleUpStatus{Result: status.ScaleUpSuccessful}, nil
+	prPods := unschedulablePods
+
+	for len(prPods) > 0 {
+		prs := provreqclient.ProvisioningRequestsForPods(o.client, prPods)
+		prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
+		if len(prs) == 0 {
+			break
+		}
+
+		// Pick 1 ProvisioningRequest.
+		pr := prs[0]
+
+		scaleUpIsSuccessful, err := o.checkcapacity(prPods, pr)
+		if err != nil {
+			scaleUpStatus, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+			combinedStatus.Add(scaleUpStatus)
+		}
+
+		if scaleUpIsSuccessful && err == nil {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpSuccessful})
+
+			if err := o.context.ClusterSnapshot.Commit(); err != nil {
+				scaleUpStatus, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+				o.context.ClusterSnapshot.Revert()
+
+				combinedStatus.Add(scaleUpStatus)
+			}
+		} else {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpNoOptionsAvailable})
+		}
+
+		provisioningRequestsProcessed++
+		if stopBatch := o.shouldStopBatchProcessing(provisioningRequestsProcessed, startTime); stopBatch {
+			break
+		}
+
+		// For batch processing, the next ProvisioningRequest's pods are fetched using the provisioningRequest pods injector.
+		// TODO: Refactor such that injector injects pods for multiple ProvisioningRequests at once. Pods should be segregated by ProvisioningRequest before being processed.
+		prPods, err = o.getNextPrPods()
+		if err != nil {
+			st, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+			combinedStatus.Add(st)
```

Comment:
Shouldn't we `break` here?

Reply:
This can return an error if the pod-template for a ProvisioningRequest is not found. It could be possible that the next PR might be able to be processed.

Reply:
Ack, can you add that explanation in a comment here? It looks like a bug at first glance.